Automattic/newspack-plugin
# fix(esp-sync): schedule second sync upon subscription reactivation, just in case

### All Submissions:






### Changes proposed in this Pull Request:



We fire an ESP contact sync whenever a Woo subscription status is updated. When a subscription undergoes automatic renewal, it rapidly changes status to `on-hold` when payment processing begins, and then to `active` once successful payment is confirmed. However, since this can happen so quickly it's possible that sync after the `active` update can happen before the `on-hold` sync completes (a sort of race condition).

Rather than trying to untangle the Woo Subscriptions logic here, the easiest thing to do is to schedule a second "backup" sync a few minutes later when a subscription status changes to `active`. This should ensure that the ESP contact is updated to `active` after a successful renewal even if the race condition occurs.

### How to test the changes in this Pull Request:

1. Check out this branch.
2. Manually renew a subscription in WP admin.
3. Confirm that the ESP contact is synced when the subscription is changed to `on-hold` status and again when the status changes to `active` (but note that it might not always happen in this order).
4. After the `active` sync completes, run `wp cron event list` and confirm that a non-repeating `newspack_scheduled_esp_sync` job is scheduled roughly two minutes in the future.
5. Wait ~2 minutes or run `wp cron event run newspack_scheduled_esp_sync` and confirm that the contact is synced to the ESP again with the correct "active" status.
6. (BONUS) Renew several subscriptions in rapid succession (within a single 2-minute window) and confirm that all subscriptions get a second sync ~2 minutes later, with the correct contact data.

### Other information:




---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208887537927842
